### ISSUE TITLE
WavPack export: rename the "Quality" and "Hybrid" options...

### ISSUE THREAD

**Issue by YZgitter** [audacity/audacity#5636]
### Your idea

1. Rename the "Quality" option to "Speed (Quality)" or even just "Speed", because in the default lossless mode there is obviously no quality variation.
The available choices there should also be renamed accordingly, for example "Fast (Low Quality)" and "Slowest (Highest Quality)", prioritizing the lossless mode. The first word would denote the encoding/decoding speed and the second (in parenthesis) would denote the quality, which is only relevant for the opt-in _lossy_ mode.

2. Rename the "Hybrid" option to "Hybrid (lossy)", to make it more evident that it's a lossy mode.
3. Rename the "Bit Rate" option to "Bit Rate (lossy)" for the same reason. Or simply just disable that drop-down menu if Hybrid Mode is unchecked.

### Problem to be solved

I believe that WavPack is mostly used for lossless encoding, because that's what it's good at. The default export in Audacity also produces a _lossless_ WV file.
Therefore, renaming the "Quality" and "Bit Rate" settings can prevent some confusion.

### Additional context

What I suggested here is just a quick fix based on the current layout, but I would actually push towards greater separation of the lossy mode, putting it out of easy reach for most users.
The WavPack lossy encoding frankly isn't that good (other lossy codecs perform better). I think its only value are the correction wvc files. That's an interesting, but very niche use case, mostly suitable for music collections, not Audacity exports.
WavPack's lossless mode, on the other hand, is something that should be promoted, because it allows floating point audio, unlike FLAC.

So you could go even further and remove all mentions of quality/bitrates and put the lossy hybrid mode in a separate (advanced) section.

**Comment by dbry** [comment]
Hi @LWinterberg, and thanks for pointing me to this thread, I would not have found it otherwise!

I agree that we don’t want anyone selecting the hybrid mode by mistake, and I understand that the hybrid options could certainly be a source of confusion among users. The idea of somehow isolating them to another (advanced) submenu. or even eliminating them, seems reasonable.

However, I’m not sure I agree that the word “Quality” is not applicable to the lossless compression mode. It doesn’t affect the _audio_ quality (obviously), but it does affect the quality of the compression (i.e., the _amount_ of compression). If the menu item only referred to speed (e.g., "Fast" or "Slow") then why would anyone select “Slow” without additional context? The “Slow” mode provides _better_ compression, but if you think that the word “quality” is confusing it could certainly be changed to “Level” or “Mode”.

I agree that WavPack’s hybrid lossless (2-file) mode is a quirky niche case, however I disagree that the hybrid lossy mode is useless. As was suggested, it does not compare well to transform codecs (like MP3, etc.) at low bitrates, but it’s not really meant to. Even lossless compression of high-resolution audio (especially at high sampling rates) is still extremely wasteful and I feel that there exists a place for an intermediate solution for distributing and archiving audio projects.

For example, at 8.0 bits/sample, WavPack’s hybrid lossy mode can compress floating-point audio with complete transparency while still being sample-accurate and retaining all the standard dynamic range and clipping immunity advantages of floating point. This will work with any sample rate, any number of channels, and always provide 4:1 compression, which is more than twice what the lossless mode would usually provide. I realize that people are not comfortable with this (e.g., “better safe than sorry”), but if you try this mode and do a invert-and-paste and look at (and listen to) the difference I think you’ll agree that it’s not an unreasonable use case (the noise is generally inaudible in isolation, _even without the masking effect of the audio!_)

Of course, I agree that nobody should accidentally get this mode, and you wouldn’t want to do it repeatedly on the same audio. When I inherited this project the hybrid modes were there (probably because I have them in my [Audition plugins](https://github.com/dbry/amioWavpack)) and I never considered removing them, but if that’s the consensus I have no problem with it. [Reaper](https://www.reaper.fm/about.php) has great support for WavPack and does not have the hybrid modes available for export, and the same hybrid operations are possible (and maybe even easier) using the WavPack command-line tools, although that’s not everyone’s cup of tea. 

I have been working toward WavPack 5.7.0 this year and one of the things I have added is native multi-threading for encoding and decoding, using either Windows threads or pthreads. This is done in `libwavpack` and should be virtually transparent to the client (other to enable it and use larger buffers), but can provide a huge speedup in import and export. Once I have this released I will experiment with it in Audacity to see if it’s worth it and provide a patch for your consideration.

Thanks again for the support of WavPack in Audacity!!

**Comment by LWinterberg** [comment]
I don't think that there's a good word for "how hard the encoder has to try to make the file size go down". For Opus we also use "Quality" here (which opus calls "Complexity"), and for FLAC we use "Level" at the moment.

If you aren't exactly sure what a "complexity", "mode" or "level" is exactly referring to, you probably won't touch it. Quality has a slight advantage here in that, if you come in with the goal of saving a high quality or lossless file, you'd naturally crank up everything that says "quality" and "bitrate" on it to achieve that, and it just so happens that this would end you up with the results you want. It does end up with the downside that if you do know that quality usually refers to something lossy, which for a lossless format is a bit confusing.

> I agree that we don’t want anyone selecting the hybrid mode by mistake, and I understand that the hybrid options could certainly be a source of confusion among users. The idea of somehow isolating them to another (advanced) submenu. or even eliminating them, seems reasonable.

I think we may have a display bug here: 
![image](https://github.com/audacity/audacity/assets/87814144/a43b6547-2639-455a-9d18-1688cd67f86e)
Bitrate should be unavailable/greyed out in cases where Hybrid Mode is disabled, right?

Other than that, I don't think there's much risk to people accidentally enabling it. In some sense, exporting a wavpack already is the advanced mode; you'll probably only use it after someone has explained the benefits of wavpack over WAV and MP3 to you.

**Comment by dbry** [comment]
One word that could describe this is **Effort**, but I'm not sure how much I like that, especially because "fast" doesn't go well with it.

I saw the bug you mention too (I got a fresh repo and built before I responded on this thread). That menu dialog code seems to be completely refactored from when I hacked on it, and the code to enable the bitrate setting only when hybrid was selected seems to have gone missing. Anyway, I hacked at it again and it seems like I fixed the issue with this simple patch, but a real C++ programmer who understands that code should check that this makes sense (because maybe it got removed for a reason).

```
diff --git a/modules/mod-wavpack/ExportWavPack.cpp b/modules/mod-wavpack/ExportWavPack.cpp
index 11412e100..dc3e2bade 100644
--- a/modules/mod-wavpack/ExportWavPack.cpp
+++ b/modules/mod-wavpack/ExportWavPack.cpp
@@ -165,6 +165,7 @@ public:
          {
             mListener->OnExportOptionChangeBegin();
             mListener->OnExportOptionChange(mOptions[OptionIDCreateCorrection]);
+            mListener->OnExportOptionChange(mOptions[OptionIDBitRate]);
             mListener->OnExportOptionChangeEnd();
          }
       }
@@ -220,9 +221,15 @@ private:
    void OnHybridModeChange(bool hybridMode)
    {
       if(hybridMode)
+      {
          mOptions[OptionIDCreateCorrection].flags &= ~ExportOption::Flags::ReadOnly;
+         mOptions[OptionIDBitRate].flags &= ~ExportOption::Flags::ReadOnly;
+      }
       else
+      {
          mOptions[OptionIDCreateCorrection].flags |= ExportOption::Flags::ReadOnly;
+         mOptions[OptionIDBitRate].flags |= ExportOption::Flags::ReadOnly;
+      }
    }
 };
```

**Comment by YZgitter** [comment]
> "Very High (smallest file)"

On second thought, "Very High (_best compression_)" would be even better.

**Comment by dbry** [comment]
@YZgitter I agree with everything you're saying. It's funny, I use Foobar2000's export menu quite often and I'm not sure I ever noticed that the labels change with the hybrid mode, but it's nice (if not strictly required).

I thought a little about having the hybrid mode check box and its associated settings moved to an advanced menu. One issue I thought of was that if the user inadvertently got the hybrid mode enabled then he might use it that way for a long time without realizing it, which would obviously be very bad. One solution to that might be to have the hybrid mode not be sticky and require re-enabling it every time (which might be annoying).

The best solution, imho, would be to just have the hybrid check box on the main menu, and have enabling that expand the menu window to expose the other items (otherwise they would be invisible). I think that's better than just having them grayed out (less intimidating to not see them at all), but I don't know how doable that is.

I also agree that WavPack's various additional features and complexity is a factor in its slower acceptance. People think "jack of all trades, master of none". For an example, think of those cheap DAPs that play a multitude of formats. They all play FLAC and APE and even TTA, but WavPack is very hard to find. And yet, WavPack is standard in FFmpeg and so it should be just a matter of recognizing the files. But having all the format options (hybrid, float, DSD, 1-file, 2-file, etc.) to test makes incorporating it more of a chore and I can just imagine developers saying "oh my, let's skip this one for now". But since FFmpeg handles WavPack and recognizes its frames via fourcc codes, some players will play WavPack files just fine if you just rename them to a supported format, like .mp3. A friend of mine plays WavPack files in his Tesla that way!

**Comment by dbry** [comment]
@YZgitter @LWinterberg Hi guys, long time!  :)

So, as I alluded to above, I have released (late, of course) WavPack 5.7.0 which includes the multithreading support. I dropped it into Audacity (which wasn’t at all easy...I had to overwrite the cached `libwavpack.so` files in the Conan folder, which I’m sure is not the recommended way) and enabled it (which is trivial) and got huge improvements in WavPack inport/export speed (like 2x – 3x, even counting Audacity’s overhead). It’s now faster importing and exporting than FLAC, which is something given FLAC’s reputation and not having to deal with floats.

I’m excited to get this in assuming it makes sense to you guys. I can make the changes such that nothing will change until the new library is detected (based on `wavpack.h`). It’s fully backward compatible, so even a binary expecting the new library will work fine with the old library, and vise versa. I just need you guys to make libwavpack 5.7.0 available upstream because testing by overwriting Conan’s cache is suboptimal.

The more I think about it, the more I’m convinced that the smartest route to reduce user confusion and simplify WavPack use is to just ditch the hybrid mode. The hybrid mode does not make sense in a day-to-day workflow (and is slightly dangerous), and because multithreaded hybrid export is not supported (except for multichannel), there's going to be a new difference in speed. It also makes it easier to come up with unambiguous wording for the “speed” modes if everything is lossless. The hybrid mode is readily available using the command-line program’s transcode function for someone who really understands what’s going on, and if people scream (which they won't) then it could come back in a second menu. Obviously in the final analysis it’s up to you guys, but that’s where I’m coming from.

Thanks!

**Comment by LWinterberg** [comment]
> The more I think about it, the more I’m convinced that the smartest route to reduce user confusion and simplify WavPack use is to just ditch the hybrid mode. 

Aye, we'll be doing that for the Audacity 4 exporter. @grliszas14 is on it as part of #8544

**Comment by orev** [comment]
Found this because I was confused about the options when researching, so I think it's still worth it to address the naming. When using a lossy encoder (audio and video), the naming conventions for compression levels are usually "bitrate" (mp3, aac, vorbis, opus), or "quality" (mp3, vorbis, x264, x265), and these are understood to be related to the level of loss the user can tolerate. Confusingly some of these also have a "compression level" parameter in ffmpeg, but I've never seen anyone refer to it that way for lossy compression.

For lossless codecs, I (and I think many people) think of them more like ZIP (GZ, XZ, etc.) algorithms specialized for audio. In these programs, the amount of compression is always referred to as "level", with ranges like "fast (least compression)" to "best (slowest)", or numbers 0-9. Being lossless, it's expected that the decompressed file should be identical to the input one.

I don't like the logic of "quality" referring to the "quality of compression", because in the codec world "quality" is already well established to be related to lossy compression. Since the Audacity export function supports both lossy and lossless, the wording is more important to reduce user confusion.

WavPack's hybrid mode definitely adds to the confusion, so if that's enabled maybe the language should change back to "quality". Or as suggested allow power users to use it by command-line if they really want it.
